File: chrome/browser/android/feed/feed_scheduler_bridge.cc
// Copyright 2018 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "chrome/browser/android/feed/feed_scheduler_bridge.h"

#include "base/android/jni_android.h"
#include "base/bind.h"
#include "base/time/time.h"
#include "chrome/browser/android/feed/feed_host_service_factory.h"
#include "chrome/browser/profiles/profile.h"
#include "chrome/browser/profiles/profile_android.h"
#include "components/feed/core/feed_host_service.h"
#include "components/feed/core/feed_scheduler_host.h"
#include "jni/FeedSchedulerBridge_jni.h"

using base::android::JavaRef;
using base::android::JavaParamRef;
using base::android::ScopedJavaGlobalRef;

namespace feed {

static jlong JNI_FeedSchedulerBridge_Init(
    JNIEnv* env,
    const JavaParamRef<jobject>& j_this,
    const JavaParamRef<jobject>& j_profile) {
  Profile* profile = ProfileAndroid::FromProfileAndroid(j_profile);
  FeedHostService* host_service =
      FeedHostServiceFactory::GetForBrowserContext(profile);
  return reinterpret_cast<intptr_t>(
      new FeedSchedulerBridge(j_this, host_service->GetSchedulerHost()));
}

FeedSchedulerBridge::FeedSchedulerBridge(const JavaRef<jobject>& j_this,
                                         FeedSchedulerHost* scheduler_host)
    : j_this_(ScopedJavaGlobalRef<jobject>(j_this)),
      scheduler_host_(scheduler_host),
      weak_factory_(this) {
  DCHECK(scheduler_host_);
  scheduler_host_->RegisterTriggerRefreshCallback(base::BindRepeating(
      &FeedSchedulerBridge::TriggerRefresh, weak_factory_.GetWeakPtr()));
}

FeedSchedulerBridge::~FeedSchedulerBridge() {}

void FeedSchedulerBridge::Destroy(JNIEnv* env, const JavaRef<jobject>& j_this) {
  delete this;
}

jint FeedSchedulerBridge::ShouldSessionRequestData(
    JNIEnv* env,
    const JavaRef<jobject>& j_this,
    const jboolean j_has_content,
    const jlong j_content_creation_date_time_ms,
    const jboolean j_has_outstanding_request) {
  return static_cast<int>(scheduler_host_->ShouldSessionRequestData(
      j_has_content, base::Time::FromJavaTime(j_content_creation_date_time_ms),
      j_has_outstanding_request));
}

void FeedSchedulerBridge::OnReceiveNewContent(
    JNIEnv* env,
    const JavaRef<jobject>& j_this,
    const jlong j_content_creation_date_time_ms) {
  scheduler_host_->OnReceiveNewContent(
      base::Time::FromJavaTime(j_content_creation_date_time_ms));
}

void FeedSchedulerBridge::OnRequestError(JNIEnv* env,
                                         const JavaRef<jobject>& j_this,
                                         const jint j_network_response_code) {
  scheduler_host_->OnRequestError(j_network_response_code);
}

void FeedSchedulerBridge::OnForegrounded(JNIEnv* env,
                                         const JavaRef<jobject>& j_this) {
  scheduler_host_->OnForegrounded();
}

void FeedSchedulerBridge::OnFixedTimer(JNIEnv* env,
                                       const JavaRef<jobject>& j_this) {
  scheduler_host_->OnFixedTimer();
}

void FeedSchedulerBridge::TriggerRefresh() {
  JNIEnv* env = base::android::AttachCurrentThread();
  Java_FeedSchedulerBridge_triggerRefresh(env, j_this_);
}

}  // namespace feed

File: chrome/android/feed/core/java/src/org/chromium/chrome/browser/feed/FeedSchedulerBridge.java
// Copyright 2018 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package org.chromium.chrome.browser.feed;

import com.google.android.libraries.feed.api.requestmanager.RequestManager;
import com.google.android.libraries.feed.host.scheduler.SchedulerApi;
import com.google.search.now.wire.feed.FeedQueryProto.FeedQuery.RequestReason;

import org.chromium.base.annotations.CalledByNative;
import org.chromium.base.annotations.JNINamespace;
import org.chromium.chrome.browser.profiles.Profile;
import org.chromium.components.feed.NativeRequestBehavior;

/**
 * Provides access to native implementations of SchedulerApi.
 */
@JNINamespace("feed")
public class FeedSchedulerBridge implements SchedulerApi {
    private long mNativeBridge;
    private RequestManager mRequestManager;

    /**
     * Creates a FeedSchedulerBridge for accessing native scheduling logic.
     *
     * @param profile Profile of the user we are rendering the Feed for.
     */
    public FeedSchedulerBridge(Profile profile) {
        mNativeBridge = nativeInit(profile);
    }

    /*
     * Cleans up native half of this bridge.
     */
    public void destroy() {
        assert mNativeBridge != 0;
        nativeDestroy(mNativeBridge);
        mNativeBridge = 0;
    }

    /*
     * Sets our copy of the RequestManager. Should be done as early as possible, as the scheduler
     * will be unable to trigger refreshes until after it has a reference to a RequestManager. When
     * this is called, it is assumed that the RequestManager is initialized and can be used.
     *
     * @param requestManager The interface that allows us make refresh requests.
     */
    public void setRequestManager(RequestManager requestManager) {
        mRequestManager = requestManager;
    }

    @Override
    public int shouldSessionRequestData(SessionManagerState sessionManagerState) {
        assert mNativeBridge != 0;

        @NativeRequestBehavior
        int nativeBehavior = nativeShouldSessionRequestData(mNativeBridge,
                sessionManagerState.hasContent, sessionManagerState.contentCreationDateTimeMs,
                sessionManagerState.hasOutstandingRequest);
        // If this breaks, it is because SchedulerApi.RequestBehavior and the NativeRequestBehavior
        // defined in feed_scheduler_host.h have diverged. If this happens during a feed DEPS roll,
        // it likely means that the native side needs to be updated. Note that this will not catch
        // new values and should handle value changes. Only removals/renames will cause compile
        // failures.
        switch (nativeBehavior) {
            case NativeRequestBehavior.REQUEST_WITH_WAIT:
                return SchedulerApi.RequestBehavior.REQUEST_WITH_WAIT;
            case NativeRequestBehavior.REQUEST_WITH_CONTENT:
                return SchedulerApi.RequestBehavior.REQUEST_WITH_CONTENT;
            case NativeRequestBehavior.REQUEST_WITH_TIMEOUT:
                return SchedulerApi.RequestBehavior.REQUEST_WITH_TIMEOUT;
            case NativeRequestBehavior.NO_REQUEST_WITH_WAIT:
                return SchedulerApi.RequestBehavior.NO_REQUEST_WITH_WAIT;
            case NativeRequestBehavior.NO_REQUEST_WITH_CONTENT:
                return SchedulerApi.RequestBehavior.NO_REQUEST_WITH_CONTENT;
            case NativeRequestBehavior.NO_REQUEST_WITH_TIMEOUT:
                return SchedulerApi.RequestBehavior.NO_REQUEST_WITH_TIMEOUT;
        }

        return SchedulerApi.RequestBehavior.UNKNOWN;
    }

    @Override
    public void onReceiveNewContent(long contentCreationDateTimeMs) {
        assert mNativeBridge != 0;
        nativeOnReceiveNewContent(mNativeBridge, contentCreationDateTimeMs);
    }

    @Override
    public void onRequestError(int networkResponseCode) {
        assert mNativeBridge != 0;
        nativeOnRequestError(mNativeBridge, networkResponseCode);
    }

    public void onForegrounded() {
        assert mNativeBridge != 0;
        nativeOnForegrounded(mNativeBridge);
    }

    public void onFixedTimer() {
        assert mNativeBridge != 0;
        nativeOnFixedTimer(mNativeBridge);
    }

    @CalledByNative
    private boolean triggerRefresh() {
        if (mRequestManager != null) {
            mRequestManager.triggerRefresh(RequestReason.SCHEDULED_REFRESH, ignored -> {});
            return true;
        }
        return false;
    }

    private native long nativeInit(Profile profile);
    private native void nativeDestroy(long nativeFeedSchedulerBridge);
    private native int nativeShouldSessionRequestData(long nativeFeedSchedulerBridge,
            boolean hasContent, long contentCreationDateTimeMs, boolean hasOutstandingRequest);
    private native void nativeOnReceiveNewContent(
            long nativeFeedSchedulerBridge, long contentCreationDateTimeMs);
    private native void nativeOnRequestError(
            long nativeFeedSchedulerBridge, int networkResponseCode);
    private native void nativeOnForegrounded(long nativeFeedSchedulerBridge);
    private native void nativeOnFixedTimer(long nativeFeedSchedulerBridge);
}
